specs/core/1_shard-data-chains.md

```diff
@@ -195,34 +176,67 @@ class ShardReceiptDelta(Container):
 
 ## Helper functions
 
-### `compute_slot_of_shard_slot`
+### Misc
+
+#### `pad`
 
 ```python
-def compute_slot_of_shard_slot(slot: ShardSlot) -> Epoch:
-    return Epoch(slot // SHARD_SLOTS_PER_BEACON_SLOT)
+def pad(x: bytes, length: uint64) -> bytes:
+    assert len(x) <= length
+    return x + b'\x00' * (length - len(x))
 ```
 
-### `compute_epoch_of_shard_slot`
+#### `compute_epoch_of_shard_slot`
 
 ```python
 def compute_epoch_of_shard_slot(slot: ShardSlot) -> Epoch:
-    return Epoch(slot // SHARD_SLOTS_PER_BEACON_SLOT // SLOTS_PER_EPOCH)
+    return compute_epoch_of_slot(compute_slot_of_shard_slot(slot))
 ```
 
-### `get_shard_period_start_epoch`
+#### `compute_period_start_epoch`
 
 ```python
-def get_shard_period_start_epoch(epoch: Epoch, lookback: int=0) -> Epoch:
+def compute_period_start_epoch(epoch: Epoch, lookback: Epoch=0) -> Epoch:
     return Epoch(epoch - (epoch % EPOCHS_PER_SHARD_PERIOD) - lookback * EPOCHS_PER_SHARD_PERIOD)
 ```
 
-### `get_period_committee`
+#### `compute_flat_shard_header`
 
 ```python
-def get_period_committee(state: BeaconState, epoch: Epoch, shard: Shard) -> Sequence[ValidatorIndex]:
+def compute_flat_shard_header(block: ShardBlock) -> Bytes[SHARD_HEADER_SIZE]:
     """
-    Return committee for a period. Used to construct persistent committees.
+    Return a flat serialisation of the ``block`` header which preserves hash tree root.
```

Comment:
>      Return a flat serialisation of the ``block`` header which preserves hash tree root. 

I don't think this rework accomplishes this objective. The SSZ tree structure that the new PR creates is as follows:

```python
        r
      /   \
    / \  / \
s bcr pr sr agg totb br pad sigr
                            / \
                         sig0 sig1
                         /\     /\
                       /\  /\ /\  /\
                      d d d 0 d d d 0
```

Whereas the old one was

```python
         r
     /         \
   body       sigs
   / \         / \
            sig0 sig1
             /\    /\
           /\  /\ /\ /\
......... d d d 0 d d d 0
```

The structuring of "core" and "signatures" was there to precisely create two trees of equal size (core has 7 elements that gets rounded up to 8, and signatures have two elements of 3 chunks each, which get rounded up to 2 * 4 = 8). Signatures being an object under the header makes the tree "imbalanced" with signatures being in a lower position.

Reply:
What about

````python
class ShardBlock(Container):
    slot: ShardSlot
    beacon_chain_root: Hash
    parent_root: Hash
    state_root: Hash
    aggregation_bits: Bitvector[TARGET_PERSISTENT_COMMITTEE_SIZE * 2]
    total_bytes: uint64
    body: Bytes[SHARD_BLOCK_SIZE_LIMIT - SHARD_HEADER_SIZE]
    padding: Bytes[32]
    attester_signature: BLSSignature
    proposer_signature: BLSSignature
```